File: CModuleSrc/stack_module.c
// SPDX-License-Identifier: GPL-2.0-only

#include <linux/init.h>
#include <linux/module.h>
#include <linux/miscdevice.h>
#include <linux/fs.h>
#include <linux/uaccess.h>
#include <linux/mutex.h>

MODULE_LICENSE("GPL");
MODULE_AUTHOR("Afif Rafiqin Adnan");
MODULE_DESCRIPTION("STACK MODULE using Misc Device + C");

#define DEVICE_NAME "stack_module_c"
#define BUFFER_SIZE 127
#define INSERT_VALUE _IOR('|', 1, int)
#define POP_VALUE _IOW('|', 2, int)

struct device_buffer {
	unsigned char buffer[BUFFER_SIZE];
	size_t pointer;
	struct mutex lock;
};

static struct device_buffer dev_buffer = { .buffer = { 0 },
					   .pointer = 0,
					   .lock = __MUTEX_INITIALIZER(
						   dev_buffer.lock) };

static long stack_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
{
	unsigned char value;
	long retval = 0;

	mutex_lock(&dev_buffer.lock);

	switch (cmd) {
	case INSERT_VALUE:
		if (copy_from_user(&value, (int __user *)arg,
				   sizeof(unsigned char))) {
			pr_err("Failed to read value from user\n");
			retval = -EFAULT;
			break;
		}

		dev_buffer.buffer[dev_buffer.pointer++] = value;
		pr_info("Value inserted: %d\n", value);
		break;

	case POP_VALUE:
		value = dev_buffer.buffer[--dev_buffer.pointer];
		if (copy_to_user((int __user *)arg, &value,
				 sizeof(unsigned char))) {
			pr_err("Failed to write value to user\n");
			retval = -EFAULT;
			break;
		}

		pr_info("Value popped: %d\n", value);
		break;

	default:
		pr_err("IOCTL command not recognized: %u\n", cmd);
		retval = -ENOTTY;
	}

	mutex_unlock(&dev_buffer.lock);
	return retval;
}

static int stack_open(struct inode *inode, struct file *file)
{
	mutex_lock(&dev_buffer.lock);
	memset(dev_buffer.buffer, 0, BUFFER_SIZE);
	dev_buffer.pointer = 0;
	mutex_unlock(&dev_buffer.lock);
	pr_info("stack device opened\n");
	return 0;
}

static int stack_release(struct inode *inode, struct file *file)
{
	pr_info("stack device released\n");
	return 0;
}

static const struct file_operations stack_fops = {
	.owner = THIS_MODULE,
	.unlocked_ioctl = stack_ioctl,
	.open = stack_open,
	.release = stack_release,
};

static struct miscdevice stack_module_device = {
	.minor = MISC_DYNAMIC_MINOR,
	.name = DEVICE_NAME,
	.fops = &stack_fops,
};

static int __init stack_init(void)
{
	int ret;

	ret = misc_register(&stack_module_device);
	if (ret) {
		pr_err("Failed to register misc device\n");
		return ret;
	}

	pr_info("stack misc device registered\n");
	return 0;
}

static void __exit stack_exit(void)
{
	misc_deregister(&stack_module_device);
	pr_info("stack misc device unregistered\n");
}

module_init(stack_init);
module_exit(stack_exit);

File: RustModuleSrc/stack_module.rs
// SPDX-License-Identifier: GPL-2.0-only

use core::pin::Pin;

use kernel::{
    c_str,
    ioctl::{_IOC_SIZE, _IOR, _IOW},
    miscdevice::{MiscDevice, MiscDeviceOptions, MiscDeviceRegistration},
    new_mutex,
    prelude::*,
    sync::Mutex,
    uaccess::{UserSlice, UserSliceReader, UserSliceWriter},
};

module! {
    type: StackModule,
    name: "stack_module_rust",
    author: "Afif Rafiqin Adnan",
    description: "STACK MODULE using Misc Device + Rust",
    license: "GPL",
}

const INSERT_VALUE: u32 = _IOR::<i32>('|' as u32, 1);
const POP_VALUE: u32 = _IOW::<i32>('|' as u32, 2);

struct StackModule {
    _miscdev: Pin<KBox<MiscDeviceRegistration<StackDevice>>>,
}

impl kernel::Module for StackModule {
    fn init(_module: &'static ThisModule) -> Result<Self> {
        pr_info!("Stack module device registered\n");

        let options = MiscDeviceOptions {
            name: c_str!("stack_module_rust"),
        };

        Ok(Self {
            _miscdev: KBox::pin_init(
                MiscDeviceRegistration::<StackDevice>::register(options),
                GFP_KERNEL,
            )?,
        })
    }
}

struct DeviceBuffer {
    buffer: [Option<u8>; 127],
    pointer: usize,
}

#[pin_data(PinnedDrop)]
struct StackDevice {
    #[pin]
    data: Mutex<DeviceBuffer>,
}

#[vtable]
impl MiscDevice for StackDevice {
    type Ptr = Pin<KBox<Self>>;

    fn open() -> Result<Pin<KBox<Self>>> {
        pr_info!("open was successful\n");

        KBox::try_pin_init(
            try_pin_init! {
                StackDevice { data <- new_mutex!(DeviceBuffer { buffer: [None;127], pointer: 0 }) }
            },
            GFP_KERNEL,
        )
    }

    fn ioctl(device: Pin<&StackDevice>, cmd: u32, arg: usize) -> Result<isize> {
        let size = _IOC_SIZE(cmd);

        match cmd {
            INSERT_VALUE => device.set_value(UserSlice::new(arg, size).reader())?,
            POP_VALUE => device.get_value(UserSlice::new(arg, size).writer())?,
            _ => {
                pr_err!("-> IOCTL not recognised: {}\n", cmd);
                return Err(ENOTTY);
            }
        };

        Ok(0)
    }
}

#[pinned_drop]
impl PinnedDrop for StackDevice {
    fn drop(self: Pin<&mut Self>) {
        pr_info!("release was successful\n");
    }
}

impl StackDevice {
    fn set_value(&self, mut reader: UserSliceReader) -> Result<isize> {
        let new_value = reader.read::<u8>()?;
        let mut guard = self.data.lock();

        // Store the pointer in a temporary variable
        let pointer = guard.pointer;
        guard.buffer[pointer] = Some(new_value);
        pr_info!("Value inserted: {new_value}\n");
        guard.pointer += 1; // Move to the next slot
        Ok(0)
    }

    fn get_value(&self, mut writer: UserSliceWriter) -> Result<isize> {
        let mut guard = self.data.lock();

        guard.pointer -= 1; // Move to the prev slot
        let pointer = guard.pointer;
        if let Some(value) = guard.buffer[pointer].take() {
            pr_info!("Value popped: {value}\n");
            writer.write::<u8>(&value)?;
            Ok(0)
        } else {
            pr_err!("-> Unexpected: No value at current pointer position\n");
            Err(EFAULT) // Fault error
        }
    }
}
